src/server/main.cpp

```diff
@@ -97,6 +101,32 @@ using namespace bridge_util;
             const auto& name = map[name##Handle]; \
             assert(name != NULL)
 
+#define NVPULL_STYPE() (remixapi_StructType) DeviceBridge::get_data()
+#define NVPULL_I() (int32_t) DeviceBridge::get_data()
+#define NVPULL_U32() (uint32_t) DeviceBridge::get_data()
+
+#define NVPULL_U64(DEST) { \
+            uint64_t* r = nullptr; \
+            uint32_t s = DeviceBridge::get_data((void**)&r); \
+            assert(s == 0 || sizeof(uint64_t) == s); \
+            (DEST) = *r; \
+            }
+
+// std::wstring --- remixapi_Path
+// make sure that TEMP does not go out of scope before calling the api function
+#define NVPULL_PATH(TEMP, DEST) { \
+            wchar_t* t = nullptr; \
+            const uint32_t len = DeviceBridge::getReaderChannel().data->pull((void**) &t) / sizeof(wchar_t); \
+            (TEMP) = std::wstring(t, len); \
+            (DEST) = (TEMP).c_str(); \
+            }
```

Comment:
When making this into a static inline helper, please have the return type be of `std::wstring`, so that string resource allocation and deallocation is safely handled. Right now, unless this is used very carefully, this may be used in such a way where a pointer to a string that is deallocated is passed around beyond its valid lifetime.

Reply:
Done! Still needs proper handling of the returned wstring so it doesn't go out of scope before calling the remix api function because remixapi_Path expects a wchar_t pointer and I've scoped the pulling of the individual structures (`server/main.cpp : Api_CreateOpaqueMaterial` for example)